[WebXR Hit Test][OpenXR] The hit test should be performed in the specified space

fujii opened the pull request:
#### adad6bed1fc2225c67ed875cfdbf1f591bd4a931
<pre>
[WebXR Hit Test][OpenXR] The hit test should be performed in the specified space
<a href="https://bugs.webkit.org/show_bug.cgi?id=303805">https://bugs.webkit.org/show_bug.cgi?id=303805</a>

Reviewed by Sergio Villar Senin.

The hit test should be performed in the specified space. Added
OpenXRCoordinator method spaceForHitTest() which returns an corresponding
XrSpace based on a passed PlatformXR::NativeOriginInformation.

* Source/WebKit/UIProcess/XR/openxr/OpenXRHitTestManager.cpp:
(WebKit::OpenXRHitTestManager::create):
(WebKit::OpenXRHitTestManager::requestHitTest):
* Source/WebKit/UIProcess/XR/openxr/OpenXRHitTestManager.h:
* Source/WebKit/UIProcess/XR/openxr/OpenXRInput.h:
(WebKit::OpenXRInput::inputSources):
* Source/WebKit/UIProcess/XR/openxr/OpenXRInputSource.h:
(WebKit::OpenXRInputSource::profiles const):
(WebKit::OpenXRInputSource::pointerSpace const):
(WebKit::OpenXRInputSource::gripSpace const):
(WebKit::OpenXRInputSource::handle const):
* Source/WebKit/UIProcess/XR/openxr/PlatformXROpenXR.cpp:
(WebKit::OpenXRCoordinator::requestHitTestSource):
(WebKit::OpenXRCoordinator::requestTransientInputHitTestSource):
(WebKit::OpenXRCoordinator::cleanupSessionAndAssociatedResources):
(WebKit::OpenXRCoordinator::spaceForHitTest const):
(WebKit::OpenXRCoordinator::populateFrameData):
(WebKit::OpenXRCoordinator::createReferenceSpacesIfNeeded):
* Source/WebKit/UIProcess/XR/openxr/PlatformXROpenXR.h:

Canonical link: <a href="https://commits.webkit.org/304450@main">https://commits.webkit.org/304450@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/ecaf16d96c591b7622f4b9e0df9eed09638781d7

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/135187 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/159/builds/7579 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/46451 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/142693 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/86962 "Built successfully") 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/155/builds/8216 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/156/builds/7428 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/103306 "Passed tests") | [✅ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/70511 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/138133 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/162/builds/5856 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/121166 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/84163 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/154/builds/5646 "Passed tests") | [✅ 🧪 api-mac-debug](https://ews-build.webkit.org/#/builders/165/builds/3252 "Passed tests") | [✅ 🛠 wpe-cairo-libwebrtc](https://ews-build.webkit.org/#/builders/166/builds/3283 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/114852 "Passed tests") | | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/145390 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/153/builds/7262 "Built successfully") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/39910 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/111684 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/160/builds/7306 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/161/builds/6084 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/112047 "Passed tests") | 
| | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/164/builds/5491 "Passed tests") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/117461 "Passed tests") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/61299 "Built successfully") | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/20903 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/158/builds/7316 "Built successfully") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/35596 "Passed tests") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/157/builds/7072 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/70867 "Built successfully") | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/163/builds/7293 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/152/builds/7173 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

svillar commented on Source/WebKit/UIProcess/XR/openxr/OpenXRHitTestManager.cpp:
```diff
@@ -31,6 +31,8 @@ OpenXRHitTestManager::OpenXRHitTestManager(XrSession session)
 {
 #if defined(XR_ANDROID_trackables)
+    if (!OpenXRExtensions::singleton().methods().xrCreateTrackableTrackerANDROID)
+        return;
```
I guess an assert fits better. We should not be using the hit test manager if the extension is not available

fujii replied:
Fixed.

svillar commented on Source/WebKit/UIProcess/XR/openxr/OpenXRInputSource.h:
```diff
@@ -42,6 +42,11 @@ class OpenXRInputSource {
     XrResult updateInteractionProfile();
 
+    const Vector<String>& profiles() const { return m_profiles; }
+    XrSpace pointerSpace() const { return m_pointerSpace; }
```
I know it's a bit of bikeshedding but I prefer aimSpace instead of pointerSpace. OpenXR spec talks about aim and grip poses (and spaces) so I think it's a better choice.

fujii replied:
Also renaming m_pointerSpace to m_aimSpace? Then, renaming m_pointerAction to m_aimAction? Anyway, I should rename them in another patch.

svillar replied:
Hmm yeah, maybe you're right and those are too many renames for this patch. Let's do it in a followup indeed

svillar commented on Source/WebKit/UIProcess/XR/openxr/PlatformXROpenXR.cpp:
```diff
@@ -927,19 +961,21 @@ PlatformXR::FrameData OpenXRCoordinator::populateFrameData(Box<RenderState> rend
-                };
-                results.append(WTFMove(result));
+        frameData.hitTestResults.add(pair.key, renderState->hitTestManager->requestHitTest(pair.value->offsetRay, spaceForHitTest(pair.value->nativeOrigin), renderState->frameState.predictedDisplayTime));
+    if (m_input) {
```
I know there are checks like this in the code, but the input is created on session start and destroyed on session end. This means that there is no reason why there should be no input. If anything we should have an assert but a call to populateFrameData with a null m_input is definitely not a valid code path.

fujii replied:
I'm going to fix it in https://github.com/WebKit/WebKit/pull/55239

svillar commented on Source/WebKit/UIProcess/XR/openxr/PlatformXROpenXR.cpp:
```diff
@@ -875,6 +876,39 @@ XrEnvironmentBlendMode OpenXRCoordinator::blendModeForSessionMode(Box<RenderStat
+            return XR_NULL_HANDLE;
+        }
+    }, [&](const PlatformXR::InputSourceSpaceInfo& inputSource) -> XrSpace {
+        if (!m_input)
```
See my other comment about m_input

fujii replied:
Fixed.

svillar commented on Source/WebKit/UIProcess/XR/openxr/OpenXRHitTestManager.h:
```diff
@@ -32,12 +32,13 @@ class OpenXRHitTestManager {
 private:
     XrSession m_session { XR_NULL_HANDLE };
-#if defined(XR_ANDROID_raycast)
+#if defined(XR_ANDROID_trackables)
```
Looks like a leftover from other patch?

fujii replied:
I forgot to do this in my previous patch. XrTrackableTrackerANDROID should be guarded by XR_ANDROID_trackables.

svillar commented on Source/WebKit/UIProcess/XR/openxr/OpenXRInput.h:
```diff
@@ -42,6 +42,8 @@ class OpenXRInput {
     Vector<PlatformXR::FrameData::InputSource> collectInputSources(const XrFrameState&, XrSpace) const;
     void updateInteractionProfile();
 
+    const Vector<UniqueRef<OpenXRInputSource>>& inputSources() { return m_inputSources; }
```
Hmm no, the idea of OpenXRInput is to aggregate the individual input sources. The OpenXRInput clients should not know nothing about the individual input sources. What we normally do is to call some method in OpenXRInput which in turn delegates in other methods from the individual input sources. Check collectInputSources() for example

fujii replied:
Hmm. I understand your idea, and tried it. But, it turned out it's not possible. The following two facts make things difficult.
* transient hit test needs input source handles for all input sources which are matching a profile.
* non-transient hit test does the hit test not only for input sources but also for ref spaces

svillar replied:
OK, I don't like it very much but we can land it now. Maybe we should refactor something in the future to avoid having this direct access to the input sources

fujii committed:
[WebXR Hit Test][OpenXR] The hit test should be performed in the specified space
https://bugs.webkit.org/show_bug.cgi?id=303805

Reviewed by Sergio Villar Senin.

The hit test should be performed in the specified space. Added
OpenXRCoordinator method spaceForHitTest() which returns an corresponding
XrSpace based on a passed PlatformXR::NativeOriginInformation.

* Source/WebKit/UIProcess/XR/openxr/OpenXRHitTestManager.cpp:
(WebKit::OpenXRHitTestManager::create):
(WebKit::OpenXRHitTestManager::requestHitTest):
* Source/WebKit/UIProcess/XR/openxr/OpenXRHitTestManager.h:
* Source/WebKit/UIProcess/XR/openxr/OpenXRInput.h:
(WebKit::OpenXRInput::inputSources):
* Source/WebKit/UIProcess/XR/openxr/OpenXRInputSource.h:
(WebKit::OpenXRInputSource::profiles const):
(WebKit::OpenXRInputSource::pointerSpace const):
(WebKit::OpenXRInputSource::gripSpace const):
(WebKit::OpenXRInputSource::handle const):
* Source/WebKit/UIProcess/XR/openxr/PlatformXROpenXR.cpp:
(WebKit::OpenXRCoordinator::requestHitTestSource):
(WebKit::OpenXRCoordinator::requestTransientInputHitTestSource):
(WebKit::OpenXRCoordinator::cleanupSessionAndAssociatedResources):
(WebKit::OpenXRCoordinator::spaceForHitTest const):
(WebKit::OpenXRCoordinator::populateFrameData):
(WebKit::OpenXRCoordinator::createReferenceSpacesIfNeeded):
* Source/WebKit/UIProcess/XR/openxr/PlatformXROpenXR.h:

Canonical link: https://commits.webkit.org/304450@main
